llvm/lib/Target/DirectX/DXILMetadata.cpp

```diff
@@ -81,6 +81,18 @@ void dxil::createShaderModelMD(Module &M) {
   Entry->addOperand(MDNode::get(Ctx, Vals));
 }
 
+void dxil::createDXILVersionMD(Module &M) {
+  Triple TT(M.getTargetTriple());
+  VersionTuple Ver = TT.getOSVersion();
```

Comment:
I think we should reconsider how we're representing the DXIL version. In DXC we've always made the DXIL version and Shader Model version lockstep and we've made a lot of assumptions about that.

I suggested in a separate chat channel that we should consider adding DXIL versions as sub-architectures.

I'm curious for feedback from @bharadwajy and @bogner.

Reply:
Apologies for any confusion caused by me posting the comment from an alternate account (@thegreatbeanz) is me.

I can't share anything about future plans for HLSL or DXIL, aside from what is already public. That said, I don't think we need to consider future plans to see the potential benefit of keeping DXIL and SM versions separate and clearly defined.

We haven't done a great job delineating shader model versions from DXIL versions in DXC, but they are different. A good example is the quadVote DXIL Op (described [here](https://microsoft.github.io/DirectX-Specs/d3d/HLSL_SM_6_7_QuadAny_QuadAll.html)). While the quadVote DXIL Op was added in DXIL 1.7, the QuadAny/QuadAll HLSL intrinsics were retroactively added back to Shader Model 6.0 because they can be lowered to different DXIL operations on older DXIL versions. We've also done this for other new functions like the `IsHelperLane` intrinsic added in [Shader Model 6.6](https://microsoft.github.io/DirectX-Specs/d3d/HLSL_ShaderModel6_6.html).

This means there isn't a 1:1 mapping between source-level intrinsics where shader model is our versioning, and DXIL operations where the DXIL version is our mechanism.

When we think about expanding DXIL targeting support to other languages, it is probably more clear to separate the Shader Model version (which is an HLSL concept) from the DXIL version (which is a code-generation concept).

Reply:
> This means there isn't a 1:1 mapping between source-level intrinsics where shader model is our versioning, and DXIL operations where the DXIL version is our mechanism.
>
> When we think about expanding DXIL targeting support to other languages, it is probably more clear to separate the Shader Model version (which is an HLSL concept) from the DXIL version (which is a code-generation concept).

This sounds wrong.  Shader Model is specifically a target constraint concept applying to the features used in the IR we pass to drivers.  DXIL version is an IR versioning concept.  In the past, if something in HLSL could be mapped to the constraints in a shader model by the compiler, then it was legal to use in that shader model.  It was never considered an HLSL concept/constraint.  We may want to put constraints on HLSL APIs in the future based on the shader model used, but that is not the core role *of* the shader model.

We also add things to the language/API in shader model time frames, but that doesn't mean that using those things at the source level always requires compilation to those shader models.  They just require a new enough compiler that has the source-level construct (like `QuaAny`/`QuadAll`/`IsHelperLane`).

In the past, they have advanced in lock-step because with each new shader model, we have needed updates to DXIL, making it necessary to advance that version, and we can't use a newer DXIL version for an older shader model, since drivers reporting support for a particular shader model are not expected to support any newer DXIL versions beyond that shader model.

Reply:
Issue https://github.com/llvm/llvm-project/issues/91388 is created to track validation of dxil version and shader model.